nemo_aligner/models/mm/stable_diffusion/megatron_sd_draftp_model.py

```diff
@@ -0,0 +1,421 @@
+        self.model._reset_sequence_parallelism_args()
+        return
+
+    def finished_generation(self):
+
+        self.model._restore_activation_checkpointing_args()
+        self.model._restore_sequence_parallelism_args()
+        return
+
+    @torch.no_grad()
+    def generate_log_images(self, latents, batch, model):
+
+        # setup default values for inference configs
+        # get autocast_dtype
+        if self.cfg.precision == "bf16":
+            autocast_dtype = torch.bfloat16
+        elif int(self.cfg.precision) == 32:
+            autocast_dtype = torch.float
+        elif int(self.cfg.precision) == 16:
+            autocast_dtype = torch.half
+        else:
+            raise ValueError('precision must be in [32, 16, "bf16"]')
+
+        with torch.no_grad(), torch.cuda.amp.autocast(
+            enabled=autocast_dtype in (torch.half, torch.bfloat16), dtype=autocast_dtype,
+        ):
+
+            sampler = sampling_utils.initialize_sampler(model, self.sampler_type.upper())
+
+            batch_size = len(batch)
+            cond, u_cond = sampling_utils.encode_prompt(
+                model.cond_stage_model, batch, self.unconditional_guidance_scale
+            )
+
+            _, intermediates = sampler.sample(
+                S=self.inference_steps,
+                conditioning=cond,
+                batch_size=batch_size,
+                shape=list(latents.shape[:1] + latents.shape[2:]),
+                verbose=False,
+                unconditional_guidance_scale=self.unconditional_guidance_scale,
+                unconditional_conditioning=u_cond,
+                eta=self.eta,
+                x_T=latents,
+                log_every_t=1,
+                truncation_steps=0,
+                return_logprobs=False,
+                return_mean_var=False,
+            )
+
+            # # transforms trajectores from a list of T+1 (b x image) tensors to a tensor of shape (b * (T+1), ...)
+            trajectories_predx0 = (
+                torch.stack(intermediates["pred_x0"], dim=0)
+                .transpose(0, 1)
+                .contiguous()
+                .view(-1, *intermediates["pred_x0"][0].shape[1:])
+            )
+
+            vae_decoder_output = []
+            idx_denoised_imgs = [self.inference_steps + (self.inference_steps + 1) * i for i in range(batch_size)]
+
+            for i in range(0, batch_size, self.vae_batch_size):
+                image = self.model.differentiable_decode_first_stage(
+                    trajectories_predx0[idx_denoised_imgs[i : i + self.vae_batch_size]]
+                )
+
+                vae_decoder_output.append(image)
+
+            vae_decoder_output = torch.cat(vae_decoder_output, dim=0)
+            vae_decoder_output = torch.clip((vae_decoder_output + 1) / 2, 0, 1) * 255.0
+
+            log_reward = [
+                self.reward_model.get_reward(
+                    vae_decoder_output[i].unsqueeze(0).detach().permute(0, 2, 3, 1), batch
```

Comment:
for permutes like this, i'd be nice to add a comment on the shape

Reply:
added # Shape [3, dim, dim] -> [1, dim, dim, 3]